crates/runtime/src/partition/mod.rs

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright 2024-2025 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use std::{any::Any, collections::HashMap, fmt, sync::Arc};
+
+use arrow_schema::SchemaRef;
+use async_trait::async_trait;
+use datafusion::{
+    catalog::{Session, TableProvider},
+    common::Constraints,
+    datasource::TableType,
+    error::DataFusionError,
+    execution::{SendableRecordBatchStream, TaskContext},
+    logical_expr::CreateExternalTable,
+    physical_plan::{DisplayAs, DisplayFormatType, ExecutionPlan, PlanProperties},
+    prelude::Expr,
+};
+
+// [`datafusion::scalar::ScalarValue`] does not necessary traits for Hash key.
+type ScalarValueString = String;
+
+#[derive(Debug)]
+pub struct PartitionTableProvider {
+    schema: SchemaRef,
+    _partition_by: Vec<Expr>,
+    _partitions: HashMap<ScalarValueString, Arc<dyn TableProvider>>,
+    _creator: PartitionCreator,
+}
+
+#[derive(Debug)]
+struct PartitionCreator {}
+
+impl PartitionTableProvider {
+    #[must_use]
+    pub fn new(_partition_by: Vec<Expr>, _cmd: &CreateExternalTable) -> Self {
```

Comment:
I don't think `CreateExternalTable` is the right interface here. As mentioned in another comment, we need to create the PolyTableProvider properly which this wouldn't. I think this should be a lambda that takes a `location` and then creates a `dyn TableProvider`. Or maybe this should be the PartitionCreator itself that is passed in? That might make more sense, if you want to create a PartitionTableProvider, you need to provide the struct that knows how to create partitions.

Reply:
And I would see if you can make the `PartitionCreator` itself a trait object instead of a struct.